src/storage/src/client/upload_object_buffered.rs

```diff
@@ -658,30 +665,83 @@ where
+                    return Ok(*o);
+                }
+                PartialUpload::Partial {
+                    persisted_size,
+                    chunk_remainder,
+                } => {
+                    offset = persisted_size;
+                    // TODO(#2043) - handle partial uploads
+                    assert_eq!(chunk_remainder, 0);
+                    remainder = r;
+                }
+            }
+        }
+    }
+
+    async fn partial_upload_request(
+        &self,
+        upload_url: &str,
+        offset: usize,
+        chunk: VecDeque<bytes::Bytes>,
+        target_size: usize,
+    ) -> Result<(reqwest::RequestBuilder, usize)> {
+        let chunk_size = chunk.iter().fold(0, |s, b| s + b.len());
+        let range = match chunk_size {
+            0 => format!("bytes */{offset}"),
+            n if n == target_size => format!("bytes {}-{}/*", offset, offset + n - 1),
+            n => format!("bytes {}-{}/{}", offset, offset + n - 1, offset + n),
+        };
         let builder = self
             .inner
             .client
             .request(reqwest::Method::PUT, upload_url)
             .header("content-type", "application/octet-stream")
-            .header("Content-Range", "bytes */*")
+            .header("Content-Range", range)
             .header(
                 "x-goog-api-client",
                 reqwest::header::HeaderValue::from_static(&self::info::X_GOOG_API_CLIENT_HEADER),
             );
 
-        let builder = apply_customer_supplied_encryption_headers(builder, self.params);
+        let builder = apply_customer_supplied_encryption_headers(builder, self.params.clone());
         let builder = self.inner.apply_auth_headers(builder).await?;
-        let stream = Box::pin(unfold(Some(self.payload), move |state| async move {
-            use crate::upload_source::StreamingSource;
+        let stream = unfold(Some(chunk), move |state| async move {
             if let Some(mut payload) = state {
-                if let Some(next) = payload.next().await {
-                    return Some((next, Some(payload)));
+                if let Some(next) = payload.pop_front() {
+                    return Some((Ok::<bytes::Bytes, Error>(next), Some(payload)));
                 }
             }
             None
-        }));
-        let builder = builder.body(reqwest::Body::wrap_stream(stream));
+        });
+        Ok((builder.body(reqwest::Body::wrap_stream(stream)), chunk_size))
+    }
+
+    async fn upload_request(mut self, upload_url: String) -> Result<reqwest::RequestBuilder> {
+        let (chunk, target_size) = {
+            let mut chunk = VecDeque::new();
+            let mut size = 0_usize;
+            while let Some(b) = self.payload.next().await.transpose().map_err(Error::io)? {
+                size += b.len();
+                chunk.push_back(b);
+            }
+            (chunk, size)
+        };
+        let target_size = target_size.div_ceil(RESUMABLE_UPLOAD_QUANTUM);
```

Comment:
wat.

Reply:
I am totally confused by what `target_size` is supposed to be. I thought `target_size == RESUMABLE_UPLOAD_QUANTUM`.

Reply:
Doh!  Okay, I removed the ugly hack.